### PR TITLE
Limit the size of served files

### PR DESCRIPTION
**Pull request by pietroalbini**
This PR limits how big the files we serve from docs.rs are, to avoid OOMs. The current limit is 5MB, but that can be changed by tweaking a constant in the source code.

This PR adds limits when downloding the files from S3 and the database, and when decompressing previously downloaded files. For now the user will see a "resource not found" error page, but a dedicated page can be created in the future.

This PR also refactors compression a bit to better enable testing. It's best to review the PR commit-by-commit.

r? @jyn514

### CODE REVIEW

File: src/utils/sized_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    pub(crate) fn reserve(&mut self, amount: usize) {
+        if self.inner.len() + amount > self.limit {
+            self.inner.reserve_exact(self.limit - self.inner.len());
+        } else {
+            self.inner.reserve_exact(amount);
+        }
+    }
```
Comment (Kixiron) [inline src/utils/sized_buffer.rs:22]: Maybe this should return the amount that it ended up reserving so that we can pivot off of that as needed, allowing us to maybe less-fatally operate by retrying writes after the buffer's flushed

File: src/storage/mod.rs [hunk 2]
```diff
@@ -324,12 +324,12 @@ mod test {
...
             );
 
-            let file = backend.get("rustdoc/Cargo.toml").unwrap();
+            let file = backend.get("rustdoc/Cargo.toml", std::usize::MAX).unwrap();
             assert_eq!(file.content, b"data");
             assert_eq!(file.mime, "text/toml");
             assert_eq!(file.path, "rustdoc/Cargo.toml");
 
-            let file = backend.get("rustdoc/src/main.rs").unwrap();
+            let file = backend.get("rustdoc/src/main.rs", std::usize::MAX).unwrap();
```
Comment (Kixiron) [inline src/storage/mod.rs:349]: I think these tests could be improved by setting the limits to under, over and exactly at the length of the content we give, that way we can ensure that the buffering works with actual files as well

File: src/web/file.rs [hunk 3]
```diff
@@ -5,12 +5,14 @@ use crate::{db, error::Result};
...
 
+const MAX_FILE_SIZE: usize = 5 * 1024 * 1024; // 5MB
+
 pub(crate) struct File(pub(crate) db::file::Blob);
 
 impl File {
     /// Gets file from database
     pub fn from_path(conn: &Connection, path: &str) -> Result<File> {
-        Ok(File(db::file::get_path(conn, path)?))
+        Ok(File(db::file::get_path(conn, path, MAX_FILE_SIZE)?))
```
Comment (Kixiron) [inline src/web/file.rs:15]: Maybe we should use an environmental variable so we can adjust the limit on the fly instead of having to redeploy

File: src/storage/mod.rs [hunk 4]
```diff
@@ -163,7 +169,8 @@ impl<'a> Storage<'a> {
                     .map(|file| (file_path, file))
             })
             .map(|(file_path, file)| -> Result<_, Error> {
-                let (content, alg) = compress(file)?;
+                let alg = DEFAULT_COMPRESSION;
+                let content = compress(file, alg)?;
```
Comment (Kixiron) [inline src/storage/mod.rs:173]: What's the purpose of setting a variable to `DEFAULT_COMPRESSION` instead of using it directly?

File: src/storage/mod.rs [hunk 5]
```diff
@@ -14,6 +14,7 @@ use std::io::Read;
 use std::path::{Path, PathBuf};
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
+const DEFAULT_COMPRESSION: CompressionAlgorithm = CompressionAlgorithm::Zstd;
```
Comment (Kixiron) [inline src/storage/mod.rs:17]: These could also be good targets for an environmental variable of some sort, we don't have multiple algos atm but in the future we might, and adjusting concurrent uploads would also be nice

File: src/storage/mod.rs [hunk 6]
```diff
@@ -14,6 +14,7 @@ use std::io::Read;
 use std::path::{Path, PathBuf};
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
```
Comment (Kixiron) [inline src/storage/mod.rs:16]: Unrelated, but it occurred to me that we could potentially benefit from dynamically selecting the number of concurrent uploads based on the current conditions, e.g. decrease the number of uploads if we're under heavier load

File: src/storage/mod.rs [hunk 7]
```diff
@@ -341,7 +349,8 @@ mod test {
     fn test_batched_uploads() {
         let uploads: Vec<_> = (0..=MAX_CONCURRENT_UPLOADS + 1)
             .map(|i| {
-                let (content, alg) = compress("fn main() {}".as_bytes()).unwrap();
+                let alg = DEFAULT_COMPRESSION;
+                let content = compress("fn main() {}".as_bytes(), alg).unwrap();
```
Comment (Kixiron) [inline src/storage/mod.rs:362]: Again it seems strange to set a variable to `DEFAULT_COMPRESSION`

File: src/storage/mod.rs [hunk 8]
```diff
@@ -369,16 +378,20 @@ mod test {
...
+            let blob = Blob {
+                mime: "text/rust".into(),
+                content: data.clone(),
+                path: "main.rs".into(),
+                date_updated: Utc::now(),
+                compression: Some(*alg),
+            };
+            test_roundtrip(std::slice::from_ref(&blob));
+            assert_eq!(decompress(data.as_slice(), *alg).unwrap(), orig.as_bytes());
+        }
```
Comment (Kixiron) [inline src/storage/mod.rs]: This could also be a good place to make sure the buffering works (ensuring it works for compressed files as well as db ones)

File: src/storage/mod.rs [hunk 9]
```diff
@@ -390,7 +399,27 @@ mod test {
...
+            let data = compress(orginal, *alg).unwrap();
+
+            // Ensure decompressing a file over the limit returns a SizeLimitReached error.
+            let err = decompress(data.as_slice(), *alg, 1000).unwrap_err();
+            assert!(err
+                .downcast_ref::<std::io::Error>()
+                .and_then(|io| io.get_ref())
+                .and_then(|err| err.downcast_ref::<crate::error::SizeLimitReached>())
+                .is_some());
         }
```
Comment (Kixiron) [inline src/storage/mod.rs]: Testing edge cases here as well would be good, like undersized and files that are on the buffer limit

File: src/storage/s3.rs [hunk 10]
```diff
@@ -190,6 +191,47 @@ pub(crate) mod tests {
         });
     }
 
+    #[test]
+    fn test_get_too_big() {
```
Comment (Kixiron) [inline src/storage/s3.rs:195]: Buffer testing here as well?

File: src/storage/database.rs [hunk 11]
```diff
@@ -16,18 +16,37 @@ impl<'a> DatabaseBackend<'a> {
         Self { conn }
     }
 
-    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+    pub(super) fn get(&self, path: &str, max_size: usize) -> Result<Blob, Error> {
         use std::convert::TryInto;
 
+        // The maximum size for a BYTEA (the type used for `content`) is 1GB, so this cast is safe:
+        // https://www.postgresql.org/message-id/162867790712200946i7ba8eb92v908ac595c0c35aee%40mail.gmail.com
+        let max_size = max_size.min(std::i32::MAX as usize) as i32;
```
Comment (Kixiron) [inline src/storage/database.rs:24]: Maybe add a debug assertion here to make sure that postgres behaves as expected

File: src/storage/database.rs [hunk 12]
```diff
@@ -91,22 +110,69 @@ mod tests {
...
             // Test that other files are not returned
             assert!(backend
-                .get("dir/bar.txt")
+                .get("dir/bar.txt", std::usize::MAX)
                 .unwrap_err()
                 .downcast_ref::<PathNotFoundError>()
                 .is_some());
             assert!(backend
-                .get("foo.txt")
+                .get("foo.txt", std::usize::MAX)
```
Comment (Kixiron) [inline src/storage/database.rs:123]: The usages of `usize::MAX` seem strange to me, it seems like testing with a sane-ish number is better and more realistic than this, which is essentially not having a limit at all

File: src/storage/database.rs [hunk 13]
```diff
@@ -91,22 +110,69 @@ mod tests {
...
                 .unwrap_err()
                 .downcast_ref::<PathNotFoundError>()
                 .is_some());
 
             Ok(())
         });
     }
+
+    #[test]
+    fn test_get_too_big() {
```
Comment (Kixiron) [inline src/storage/database.rs:133]: I know I've said it a lot, but maybe more buffering tests here?

File: src/storage/database.rs (hunk 11 above)
Comment (pietroalbini) [reply src/storage/database.rs:24]: Hmm, I don't see a way of asserting what the postgres storage limit is. Do you have any idea that doesn't involve writing 1GB of data in the database?

File: src/storage/mod.rs (hunk 4 above)
Comment (pietroalbini) [reply src/storage/mod.rs:173]: I put `DEFAULT_COMPRESSION` in a variable as that makes changing the function to accept the algorithm as a parameter easier in the future. I'm not too attached to that though, I can surely change this.

File: src/storage/mod.rs (hunk 7 above)
Comment (pietroalbini) [reply src/storage/mod.rs:362]: See the other comment.

File: src/storage/mod.rs (hunk 2 above)
Comment (pietroalbini) [reply src/storage/mod.rs:349]: Hmm, these tests are not testing the size limits, they're just the older tests I had to change due to the new argument.

File: src/storage/s3.rs (hunk 10 above)
Comment (pietroalbini) [reply src/storage/s3.rs:195]: What do you mean?

File: src/utils/sized_buffer.rs (hunk 1 above)
Comment (pietroalbini) [reply src/utils/sized_buffer.rs:22]: Do you see this being needed anywhere in the current PR? If future code needs it, the return value can be added in a future PR.

File: src/storage/database.rs (hunk 12 above)
Comment (pietroalbini) [reply src/storage/database.rs:123]: These tests are not testing the size limit, so there is no point in setting a lower limit.

File: src/storage/database.rs (hunk 13 above)
Comment (pietroalbini) [reply src/storage/database.rs:133]: What do you mean?

File: src/storage/mod.rs (hunk 5 above)
Comment (pietroalbini) [reply src/storage/mod.rs:17]: I don't think that's required for the current PR. Feel free to send a future PR though.

File: src/web/file.rs (hunk 3 above)
Comment (pietroalbini) [reply src/web/file.rs:15]: Hmm, I don't think an environment variable is strictly needed, and implementing it correctly might be complex (I'd prefer not to read and parse it every single request).

File: src/storage/mod.rs (hunk 8 above)
Comment (pietroalbini) [reply src/storage/mod.rs]: Hmm? This is not testing buffering.

File: src/storage/mod.rs (hunk 9 above)
Comment (pietroalbini) [reply src/storage/mod.rs]: Good point.

File: src/storage/mod.rs (hunk 5 above)
Comment (Nemo157) [reply src/storage/mod.rs:17]: With custom dictionaries the compression will probably be chosen on a per-file basis, so not easily configurable other than by changing code.

**Comment by Nemo157** [comment]
LGTM

File: src/storage/mod.rs (hunk 8 above)
Comment (Kixiron) [reply src/storage/mod.rs]: Buffering is still involved within the pipeline though, isn't it?

File: src/storage/database.rs (hunk 11 above)
Comment (Kixiron) [reply src/storage/database.rs:24]: I meant asserting that the size we're given matches what we believe will happen, e.g `assert!(max_size <= i32::max_value() as usize);`

File: src/storage/s3.rs (hunk 10 above)
Comment (Kixiron) [reply src/storage/s3.rs:195]: Well, the sized buffer is used indirectly by the things being tested, and so making sure that the sized buffer preforms as expected in a real scenario seems logical to me

File: src/storage/mod.rs (hunk 8 above)
Comment (pietroalbini) [reply src/storage/mod.rs]: Yeah, but the fact buffering works is already tested elsewhere. We don't need to add such tests every time we access the database.

File: src/storage/database.rs (hunk 11 above)
Comment (pietroalbini) [reply src/storage/database.rs:24]: Well, the line we're commenting on already ensures that... Also, the tests will catch it if for whatever reason it stops working (I needed to add this line *because* tests were failing).

**Comment by pietroalbini** [comment]
Addressed all the review comments, and changed the maximum size to 50MB for non-HTML files.

I think this is ready to go, I'll follow up with a PR that does a couple refactorings and then adds a test for `File::from_path`. One last check @Kixiron @jyn514?

**Comment by pietroalbini** [comment]
Nevermind, also added the final test.

File: src/storage/database.rs [hunk 14]
```diff
@@ -26,29 +26,28 @@ impl<'a> DatabaseBackend<'a> {
...
-                Err(
-                    std::io::Error::new(std::io::ErrorKind::Other, crate::error::SizeLimitReached)
-                        .into(),
-                )
-            }
+            Err(PathNotFoundError.into())
         } else {
             let row = rows.get(0);
+
+            if row.get::<_, bool>("is_too_big") {
```
Comment (jyn514) [inline src/storage/database.rs:43]: nit: I think rust can infer the type here.
```suggestion
            if row.get("is_too_big") {
```

File: src/storage/mod.rs [hunk 15]
```diff
@@ -369,16 +387,60 @@ mod test {
...
+        const MAX_SIZE: usize = 1024;
+
+        let small = &[b'A'; MAX_SIZE / 2] as &[u8];
+        let exact = &[b'A'; MAX_SIZE] as &[u8];
+        let big = &[b'A'; MAX_SIZE * 2] as &[u8];
+
+        for alg in CompressionAlgorithm::AVAILABLE {
+            let compressed_small = compress(small, *alg).unwrap();
+            let compressed_exact = compress(exact, *alg).unwrap();
+            let compressed_big = compress(big, *alg).unwrap();
```
Comment (jyn514) [inline src/storage/mod.rs:420]: nit: you don't need all this `*alg`:
```suggestion
        for &alg in CompressionAlgorithm::AVAILABLE {
            let compressed_small = compress(small, alg).unwrap();
            let compressed_exact = compress(exact, alg).unwrap();
            let compressed_big = compress(big, alg).unwrap();
```

File: src/bin/cratesfyi.rs [hunk 16]
```diff
@@ -1,17 +1,20 @@
...
 use structopt::StructOpt;
 
-pub fn main() {
+pub fn main() -> Result<(), Error> {
     let _ = dotenv::dotenv();
     logger_init();
 
-    CommandLine::from_args().handle_args();
+    CommandLine::from_args().handle_args()?;
+    Ok(())
```
Comment (jyn514) [inline src/bin/cratesfyi.rs:17]: nit: I think this looks nicer.
```suggestion
    CommandLine::from_args().handle_args()
```

File: src/config.rs [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct Config {
+    pub(crate) max_file_size: usize,
+    pub(crate) max_file_size_html: usize,
+}
+
+impl Config {
+    pub fn from_env() -> Result<Self, Error> {
```
Comment (jyn514) [inline src/config.rs:13]: The idea is that over time we'll move more of the environment variables here, instead of being spread out over the codebase? I like it, it makes things much easier to test as well.

**Comment by jyn514** [comment]
Looks good to me modulo nits.

File: src/config.rs (hunk 17 above)
Comment (pietroalbini) [reply src/config.rs:13]: Yep, that's the goal!

**Comment by pietroalbini** [comment]
@jyn514 addressed the nits.

I also updated how tests can override configuration values to this:

```rust
env.override_config(|config| {
   config.key = value;
});
```

Using the closure allows to change the default configuration for tests in the future without having to update the base config in each test that overrides it.

File: src/config.rs (hunk 17 above)
Comment (Kixiron) [reply src/config.rs:13]: I completely support that, having a concrete storage of env vars in one place would be fantastic

**Comment by jyn514** [comment]
Looks great, thanks!
